internal/util/util.go

```diff
@@ -38,22 +40,73 @@ func BundleInstanceProvisionerFilter(provisionerClassName string) predicate.Pred
 func MapBundleToBundleInstanceHandler(cl client.Client, log logr.Logger) handler.MapFunc {
 	return func(object client.Object) []reconcile.Request {
 		b := object.(*rukpakv1alpha1.Bundle)
+
 		bundleInstances := &rukpakv1alpha1.BundleInstanceList{}
-		var requests []reconcile.Request
 		if err := cl.List(context.Background(), bundleInstances); err != nil {
 			log.WithName("mapBundleToBundleInstanceHandler").Error(err, "list bundles")
-			return requests
+			return nil
 		}
+		var requests []reconcile.Request
 		for _, bi := range bundleInstances.Items {
 			bi := bi
-			if bi.Spec.BundleName == b.Name {
-				requests = append(requests, reconcile.Request{NamespacedName: client.ObjectKeyFromObject(&bi)})
+			for _, ref := range b.GetOwnerReferences() {
+				if ref.Name == bi.GetName() {
+					requests = append(requests, reconcile.Request{NamespacedName: client.ObjectKeyFromObject(&bi)})
+				}
 			}
 		}
 		return requests
 	}
 }
 
+// GetBundlesForBundleInstanceSelector is responsible for returning a list of
+// Bundle resource that exist on cluster that match the label selector specified
+// in the BI parameter's spec.Selector field.
+func GetBundlesForBundleInstanceSelector(ctx context.Context, c client.Client, bi *rukpakv1alpha1.BundleInstance) ([]*rukpakv1alpha1.Bundle, error) {
+	bundleSelector, err := metav1.LabelSelectorAsSelector(bi.Spec.Selector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse the %s label selector: %w", bi.Spec.Selector.String(), err)
+	}
+	bundleList := &rukpakv1alpha1.BundleList{}
+	if err := c.List(ctx, bundleList, &client.ListOptions{
+		LabelSelector: bundleSelector,
+	}); err != nil {
+		return nil, fmt.Errorf("failed to list bundles using the %s selector: %w", bundleSelector.String(), err)
+	}
+	bundles := []*rukpakv1alpha1.Bundle{}
+	for _, b := range bundleList.Items {
+		bundles = append(bundles, b.DeepCopy())
+	}
+	return bundles, nil
+}
+
+// CheckExistingBundlesMatchesTemplate evaluates whether the existing list of Bundle objects
+// match the desired Bundle template that's specified in a BundleInstance object. If a match
+// is found, that Bundle object is returned, so callers are responsible for nil checking the result.
+func CheckExistingBundlesMatchesTemplate(existingBundles []*rukpakv1alpha1.Bundle, desiredBundleTemplate *rukpakv1alpha1.BundleTemplate) *rukpakv1alpha1.Bundle {
+	for _, bundle := range existingBundles {
+		if !CheckDesiredBundleTemplate(bundle, desiredBundleTemplate) {
+			continue
+		}
+		return bundle
+	}
+	return nil
+}
+
+// CheckDesiredBundleTemplate is responsible for determining whether the existingBundle
+// parameter is semantically equal to the desiredBundle Bundle template.
+func CheckDesiredBundleTemplate(existingBundle *rukpakv1alpha1.Bundle, desiredBundle *rukpakv1alpha1.BundleTemplate) bool {
```

Comment:
nit: does this function have to be exported if its only called in this package in the above function?

Reply:
I think you were looking at an older version of this PR as I've since changed it to be called in tests as well.